mu/virtual_environment.py

```diff
@@ -487,15 +530,53 @@ def ensure_and_create(self, emitter=None):
         while True:
             logger.debug("Checking virtual environment; attempt #%d.", n)
             try:
+                #
+                # try_to_create will be True if this is a brand new install
+                # or if we've failed to ensure at least once
+                #
                 if try_to_create:
                     self.create()
+                    try_to_create = False
+
+                #
+                # Otherwise check whether the venv settings match the version of Mu
+                # If not, recreate the venv with updated baseline packages and
+                # the existing user packages
+                #
+                else:
+                    venv_mu_version = self.settings.get(
+                        "mu_version", "-no-version-"
+                    )
+                    if venv_mu_version != mu_version:
+                        logger.warning(
+                            "Venv created by Mu version %s; Current Mu is version %s",
+                            venv_mu_version,
+                            mu_version,
+                        )
+                        self.recreate()
+
+                #
+                # In any situation (initial creation, recreation after Mu update,
+                # regular run) ensure that the venv is still valid
+                #
                 self.ensure()
                 logger.info("Valid virtual environment found at %s", self.path)
+
+                #
+                # If we reach this point, ensure hasn't raised an exception and
+                # we're good to save settings and move on
+                #
                 self.settings.save()
                 break
 
-            except VirtualEnvironmentError:
-                self.quarantine_failed_venv()
+            except VirtualEnvironmentError as exc:
+                #
+                # If any of the ensure/create steps cause a VirtualEnvironmentError
+                # retry up to a maximum number in case we've just been unlucky
+                # with network, file contention etc.
+                #
+                logger.error(exc.message)
+                self.quarantine_venv()
                 if n < n_tries:
                     self.relocate(self._generate_dirpath())
                     try_to_create = True

```

Comment:
Any reason not to use `self.recreate()` here?

Reply:
I think I was going for a broad symmetry around the loop, but I will take a look and see if a `recreate` makes more sense.